### ISSUE TITLE
An issue in the header of wav demuxed from blu-ray m2ts files 

### ISSUE THREAD

**Issue by rena-minami** [justdan96/tsMuxer#272]
When I encode the wav demuxed from a bluray pcm file, the flac encoder reports `ERROR: got partial sample`. While using the same track demuxed with eac3to, the encoder can proceed flawlessly.  
Here are the differences highlighted:
<img width="484" alt="Screen Shot 2020-04-05 at 1 24 53 AM" src="https://user-images.githubusercontent.com/63141763/78468406-6e623000-76dd-11ea-9486-1525e0a589d4.png">
<img width="483" alt="Screen Shot 2020-04-05 at 1 25 04 AM" src="https://user-images.githubusercontent.com/63141763/78468408-77eb9800-76dd-11ea-85f8-131dd29577c3.png">
It seems that the "file size" of the header in tsMuxer is filled with `FFFFFF` for both file size (4 bytes after "RIFF") and data size (4 bytes after "data"), (this behavior seems to be hardcoded in line 638 and 655 in `lpcmSteamReader.cpp`). And if I replace the latter `FFFFFF` with the actual size, which is `A0DEC218`, the flac encoder works just fine. 

Meanwhile, the real values for the two `FFFFFF`s, are somehow written at the end of the WAV file, causing it having one more sample than expected. 

If there is any extra information needed, please reply and I will provide accordingly. Thanks.

**Comment by justdan96** [comment]
@jcdr428 can you reproduce this issue?

**Comment by jcdr428** [comment]
@rena-minami the "ffffffff" is normaly replaced at end of reading with file size in method `beforeFileCloseEvent(File& file)` on line 721. It is working as it should on my W64. If you can debug / trace on your system, please see what happens when in this method.

@justdan96 however this has allowed me to discover a new bug in the lpcm to wav, I'll push the fix.

**Comment by rena-minami** [comment]
@jcdr428 Thanks for the reply. After some testing, I can confirm that, the wav extracted on windows is perfectly correct, however, on macOS (10.15.4) and ubuntu (18.04), the issue can be reproduced. I will try to see what happened exactly later today. 
FYI, I found the prebuilt linux binaries does not lunch (directly returns silently), I had to build from source for both tsMuxeR and the GUI.

**Comment by jcdr428** [comment]
@rena-minami so the issue must come from file_linux.cpp . I can't debug linux on my side.

**Comment by rena-minami** [comment]
@jcdr428 I have found what's happening. 

First, in `singleFileMuxer.cpp`, line 205, the file is opened with  flag `File::ofAppend`. This, in `file_unix.cpp`, adds `O_APPEND` to the system call `open`. When `open` is given such a flag, it always writes to the end of the file no matter where the current seek position is. So, the seek calls in `beforeFileCloseEvent` is successful, but useless in this case, and consequently, these file size information is written to the end of the file instead of the correct position. One can confirm this by printing the current position before and after the system call `write`, something like: before, position=4, after, position=<large number> will be observed. 

However, simply removing the `O_APPEND` flag breaks the program. So a more complicated fix maybe needed. For example, close the file first and reopen the file without `O_APPEND`. Since I am not familiar enough with the project, I would kindly ask you to make the change to the design. Thanks.

**Comment by jcdr428** [comment]
@rena-minami good catch. Please add the two lines
```
        streamInfo->m_file.close();
        streamInfo->m_file.open(streamInfo->m_fileName.c_str(), File::ofWrite);
```
before the line 207
```
        lpcmReader->beforeFileCloseEvent(streamInfo->m_file);
```
It does not break anything on Windows, so if you confirm it works for Linux and Mac, then I will push the commit.

**Comment by rena-minami** [comment]
@jcdr428 
I made a mistake, it is not the place around line 207 that caused the bug, the code invoked the `beforeFileCloseEvent` function is actually line 296 in this case. But I think that both places need fixes, since essentially they have the exact same issue.

Also, due to the logic in `file_unix.cpp`, it automatically adds `O_TRUNC` to the flags by default which deletes the whole file if exist upon opening (if `O_APPEND` is added to the flag, this is removed automatically so in line 205 and 291 there is no need to specify), but in this fix, we need to specify the flag otherwise the demuxed file will be erased. So my modification of your proposed fix is:

add
```c
        streamInfo->m_file.close();
        streamInfo->m_file.open(streamInfo->m_fileName.c_str(), File::ofWrite+File::ofNoTruncate);
```
before the function call in line 207 and 296. In the latter part, adding a pair of braces may be needed.
I have tested this fix on my systems, and it fixes the bug.
